### ISSUE TITLE
Discussion / sub-epochs or something?

### ISSUE THREAD

**Issue by danpovey** [lhotse-speech/lhotse#627]
There are situations where it might make sense to break an epoch down into smaller units.
For example, in icefall, our model-averaging approach won't work well with super-large datasets, as one epoch
will be too long.
Also, checkpointing is unfortunately not that helpful when your job dies before the first epoch finishes.
I just want to open up a discussion about this.  I don't really have any bright ideas for how to deal with this.

Another issue which I'm not sure whether it's related to this is, it would be nice to know approximately how far through an epoch we are, for purposes of things like setting the learning rate (assume we might want the learning rate schedule to know how far we are into training).

**Comment by danpovey** [comment]
... for example, a progress-indicator would make it possible to do something like:
```
  if (math.floor(sampler.proportion_done()) * params.num_sub_epochs > cur_sub_epoch):
     cur_sub_epoch += 1
     write_checkpoint(params.epoch, params.num_sub_epochs, cur_sub_epoch, ....)
```

**Comment by csukuangfj** [comment]
How about the following idea?

Add the following option to the command-line arguments:

```
Name: --save-every-iter
Type: int
```
**Description**: Save the following information when
`params.global_batch_idx % params.save_every_iter` == 0

  (1) Model parameters
  (2) Sampler's state_dict
  (3) Optimizer's state_dict

The filename can have the form f'checkpoint-iter-{params.global_batch_idx}.pt'.
To save disk space, we can limit the number of checkpoints on the disk.
For example, keep only the last N checkpoints and delete other checkpoints.

Something to pay attention to:

  The sampler does not save the number of batches processed so far in the
  current epoch. We need to note down the number of processed batches for the
  current epoch while saving the checkpoint. Suppose it is saved into a variable
  'processed_batches_this_epoch'

  When we load the checkpoint, we can use

    for batch_idx, batch in enumerate(train_dl):
      if batch_idx < params.processed_batches_this_epoch:
        continue
      params.processed_batches_this_epoch += 1
 

As for model averaging in decoding, we can choose the number of checkpoints
to average. For instance, to average K checkpoints, we use the last K checkpoints.
This is possible since the checkpoint filename contains `params.global_batch_idx`
and we can sort the filenames to select K files that have the top-K `params.global_batch_idx`.

**Comment by danpovey** [comment]
Hm...
Let's see if Piotr responds first.
It would be nice if there were a way to avoid reading data from disk in this scenario, but also I don't want to unnecessarily complicate the interfaces of Lhotse.
An advantage of the proportion_done() approach would be that it would make it much easier to implement
learning-rate schedules that are specified by epoch.
I don't much like the idea of automatically deleting all but the last K checkpoints, it would make it harder to
do diagnostics and to test intermediate iters.  
Also, if you happen to write out a model just before the end of the epoch, you would be averaging two very similar models (that, and the one at the start of the next epoch).  I know this likely isn't a big deal, but the proportion_done() approach does prevent this.
However, it may turn out that proportion_done() is difficult to implement for some reason.

**Comment by pzelasko** [comment]
My belief is that for large datasets we should never care about the number of epochs, and only care about the number of iterations. It makes the checkpointing and validation logic much simpler and independent of the dataset size so the recipes are also easier to re-use. This can be implemented using what @csukuangfj proposes.

The issue with `proportion_done` is that you need to know (or have a good estimate) up-front about how much data there is. But you might want to dynamically filter it out, or read dynamically from some very large file / network location, or maybe dynamically cut into shorter cuts, etc. Every now and then there comes some use case which makes the assumption of knowing how much data you have invalid.

Regarding restoring sampler's state: some Lhotse sampler support it (with `.state_dict()` and `.load_state_dict()`) so that you don't need to load the data to skip some part of it. The two most interesting samplers (`DynamicCutSampler` and `DynamicBucketingSampler`) don't support it yet; I plan to add the support for that later. The workflow is described in https://lhotse.readthedocs.io/en/latest/datasets.html#restoring-sampler-s-state-continuing-the-training

**Comment by danpovey** [comment]
Hm, OK.
Regarding the state dict, how much info does it have to write?
I wonder if this functionality might be hard to maintain?

**Comment by pzelasko** [comment]
The current implementations are very lightweight, they basically write down how many cuts were iterated over, and the arguments of the sampler. During loading, the sampler iterates over the input cuts to reach the same state, which is very fast with in-memory cutsets. 

For "dynamic" variants, I think we can do the same thing, although iterating over very large cutsets might take some seconds (or minutes..). Another idea is to modify the existing code so that we can pickle iterators over files / other sources and keep the byte offset of the original file where they finished reading, and during loading use `tell` to jump directly to that offset. It will then also require to store some data in the checkpoint that is present in the sampler buffers (dynamic shuffling buffer, dynamic bucketing buffer, etc.). With typical configurations that seem to work for me at this time, this might be storing sth between 20k - 100k cuts in each checkpoint, but I don't think it's too excessive and will disappear compared to parameters + optimizer states anyway. Implementing this might be a bit more involved though.

**Comment by danpovey** [comment]
I think pickling iterators etc., sounds like a can of worms, remembering the number of cuts sounds easier.

**Comment by GoVivace** [comment]
When recycle=True flag is available in the Zip and Mux samplers, the "epoch" will lose its traditional meaning, because different datasets will be at different epochs through the iterations anyways.
